cub/cub/block/block_load_to_shared.cuh

```diff
@@ -0,0 +1,395 @@
+    NV_DISPATCH_TARGET(
+      NV_PROVIDES_SM_90,
+      (return ::cuda::ptx::mbarrier_try_wait_parity(&temp_storage.bar, phase_parity);),
+      NV_PROVIDES_SM_80,
+      (const bool done = ::cuda::ptx::mbarrier_test_wait_parity(&temp_storage.bar, phase_parity); //
+       if (!done) { __nanosleep(0); } //
+       return done;),
+      NV_IS_DEVICE,
+      (__syncthreads(); //
+       return true;));
+  }
+
+public:
+  /// @smemstorage{BlockLoadToShared}
+  struct TempStorage : cub::Uninitialized<_TempStorage>
+  {};
+
+  //! @name Collective constructors
+  //! @{
+
+  //! @brief Collective constructor using a private static allocation of shared memory as temporary storage.
+  _CCCL_DEVICE _CCCL_FORCEINLINE BlockLoadToShared()
+      : temp_storage(PrivateStorage())
+  {
+    InitBarrier();
+  }
+
+  //! @brief Collective constructor using the specified memory allocation as temporary storage.
+  //!
+  //! @param[in] temp_storage
+  //!   Reference to memory allocation having layout type TempStorage
+  _CCCL_DEVICE _CCCL_FORCEINLINE BlockLoadToShared(TempStorage& temp_storage)
+      : temp_storage(temp_storage.Alias())
+  {
+    _CCCL_ASSERT(::cuda::device::is_object_from(temp_storage, ::cuda::device::address_space::shared),
+                 "temp_storage has to be in shared memory");
+    InitBarrier();
+  }
+
+  //! @}  end member group
+
+  //! @brief Collective destructor invalidates underlying @c mbarrier enabling reuse of its temporary storage.
+  //! @note
+  //! Block-synchronization is needed afterwards to reuse the shared memory from the temporary storage.
+  _CCCL_DEVICE _CCCL_FORCEINLINE ~BlockLoadToShared()
+  {
+    if (elected)
+    {
+      NV_IF_TARGET(NV_PROVIDES_SM_80,
+                   (
+                     // Stolen from cuda::barrier
+                     asm volatile("mbarrier.inval.shared.b64 [%0];" ::"r"(static_cast<::cuda::std::uint32_t>(
+                       ::__cvta_generic_to_shared(&temp_storage.bar))) : "memory");));
+    }
+  }
+
+  //! @brief Copy elements from global to shared memory
+  //! @tparam T
+  //!   **[inferred]** Value type for this transaction
+  //! @tparam GmemAlign
+  //!   Guaranteed alignment in bytes of the source range (both begin and end) in global memory
+  //! @param[in] smem_dst
+  //!   Destination buffer in shared memory that is aligned to `SharedBufferAlignBytes<T>()` and at least
+  //!   `SharedBufferSizeBytes<T, GmemAlign>(size(gmem_src))` big.
+  //! @param[in] gmem_src
+  //!   Source range in global memory, determines the size of the transaction
+  //! @return
+  //!   The range in shared memory (same size as `gmem_src`) which should be used to access the data after `Commit` and
+  //!   `Wait`.
+  // TODO Allow spans with static sizes?
+  template <typename T, int GmemAlign = alignof(T)>
+  [[nodiscard]] _CCCL_DEVICE _CCCL_FORCEINLINE ::cuda::std::span<T>
+  CopyAsync(::cuda::std::span<char> smem_dst, ::cuda::std::span<const T> gmem_src)
```

Comment:
```suggestion
  CopyAsync(::cuda::std::span<char> smem_dst, ::cuda::std::span<T> gmem_src)
```
This came up in #5926. While the `const` is correct, it is horrible for usability as `T` wont be inferred when the user is passing a modifiable `span<type> src_span`. Instead the user currently needs to call `load2sh_obj.template CopyAsync<type>(dst_span, src_span)`.